### PR TITLE
Initialize Types when calling a generic function

### PR DESCRIPTION
**Pull request by jp1729**
When calling a generic function, initialize the type of the function, as well as save information about the type and its source location in DeclRefExpr. The type of DeclRefExpr will be the completely initialized type. Any parent node will keep the initialized type.

### CODE REVIEW

File: include/clang/AST/Expr.h [hunk 1]
```diff
@@ -955,9 +955,43 @@ class DeclRefExpr final
       private llvm::TrailingObjects<DeclRefExpr, NestedNameSpecifierLoc,
                                     NamedDecl *, ASTTemplateKWAndArgsInfo,
                                     TemplateArgumentLoc> {
+public :
+  /// \brief extra information for DeclRefExprInfo.
+  class GenericFunctionCallInfo {
```
Comment (dtarditi) [inline include/clang/AST/Expr.h]: I suggest renaming this to GenericInstInfo.  Please add a comment saying that this is information for an instantiation of a generic function.   I suggested renaming TypeNameInfo to TypeArgument.

File: include/clang/AST/Expr.h [hunk 2]
```diff
@@ -955,9 +955,43 @@ class DeclRefExpr final
...
+  public:
+    struct TypeNameInfo {
+      QualType typeName;
+      TypeSourceInfo *sourceInfo;
+    };
+
+  private :
+    /// \brief If DeclRefExpr references generic function, it must store
```
Comment (dtarditi) [inline include/clang/AST/Expr.h]: I suggest rewording the comment to be more precise: references to generic functions in code must always be instantiated (applied to type arguments)  We store the type arguments in DeclRefExpr.

Instead of calling the member TypeNameInfos, I suggest TypeArguments.

File: include/clang/AST/Expr.h [hunk 3]
```diff
@@ -955,9 +955,43 @@ class DeclRefExpr final
...
+    MutableArrayRef<TypeNameInfo> typeNameInfos() {
+      return{ TypeNameInfos, NumTypeNameInfo };
+    }
+  };
+
+private :
   /// \brief The declaration that we are referencing.
   ValueDecl *D;
 
+  GenericFunctionCallInfo *GenFuncInfo;
```
Comment (dtarditi) [inline include/clang/AST/Expr.h]: I suggest renaming this to something like TypeArguments.   Please add a comment describing this field.

File: lib/Parse/ParseExpr.cpp [hunk 4]
```diff
@@ -3023,6 +3022,74 @@ ExprResult Parser::ParseBoundsExpression() {
...
 }
 
+// Recurse through types and substitute any TypedefType with TypeVariableType
+// as the underlying type.
+QualType Parser::SubstituteTypeVariable(QualType QT,
+  SmallVector<DeclRefExpr::GenericFunctionCallInfo::TypeNameInfo, 16> &typeNames) {
+  const Type *T = QT.getTypePtr();
+  switch (T->getTypeClass()) {
+  case Type::Pointer: {
+    const PointerType *pType = dyn_cast<PointerType>(T);
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: We try to practice "defensive programming" in our changes in clang, more so than usually seen. After a dyn_cast, we usually check that the result is non-null as expected.   If it is null, we use llvm_unreachable and return some sort of sensible type.   The other uses of dyn_cast in this function should be updated too.

File: lib/Parse/ParseExpr.cpp [hunk 5]
```diff
@@ -3023,6 +3022,74 @@ ExprResult Parser::ParseBoundsExpression() {
...
+    // If underlying type is TypeVariableType, must replace the entire
+    // TypedefType. If underlying type is not TypeVariableType, there cannot be
+    // a type variable type that this context can replace.
+    const TypedefType *tdType = dyn_cast<TypedefType>(T);
+    const Type *underlying = tdType->getDecl()->getUnderlyingType().getTypePtr();
+    if (const TypeVariableType *tvType = dyn_cast<TypeVariableType>(underlying))
+      return typeNames[tvType->GetIndex()].typeName;
+    else
+      return QT;
+  }
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: It is possible for various transformations in clang to eliminate the typedef.  Please add a case for type variable by itself.    Also, please add a case for function types.  Function types could use a type variable or a type that contains a type variable.

File: test/CheckedC/generic-function/generic-func-call-info.c [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Test to make sure that the generic function calls are correctly instantiated
+//
+// In CallExpr, the type of the DeclRefExpr should be completely instantiated,
+// by replacing all type variable types with specified type names in the list
+// of type names provided when calling generic function.
+//
+//
+// RUN: %clang_cc1 -ast-dump -fcheckedc-extension %s | FileCheck %s
```
Comment (dtarditi) [inline test/CheckedC/generic-function/generic-func-call-info.c]: Please add test cases for the different forms of constructed types:
- Pointers to type variables.
- Function types that have type variables within them.
- Array types that use type variables.

Also please add a test case for the new error message that you have introduced.

Comment (jp1729) [inline test/CheckedC/generic-function/generic-func-call-info.c]: I added the test case for the new error message in Checkedc repository.

**Comment by jp1729** [comment]
It's ready for further review.

File: lib/Parse/ParseExpr.cpp [hunk 7]
```diff
@@ -3023,6 +3022,142 @@ ExprResult Parser::ParseBoundsExpression() {
...
+    // TypedefType. If underlying type is not TypeVariableType, there cannot be
+    // a type variable type that this context can replace.
+    const TypedefType *tdType = dyn_cast<TypedefType>(T);
+    if (!tdType) {
+      llvm_unreachable("Type cannot be empty when substituting type variable.");
+      return QT;
+    }
+    const Type *underlying = tdType->getDecl()->getUnderlyingType().getTypePtr();
+    if (const TypeVariableType *tvType = dyn_cast<TypeVariableType>(underlying)) {
+      if (!tvType) {
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: You can remove this if-statement: we know tvType will be non-null.

File: lib/Parse/ParseExpr.cpp [hunk 8]
```diff
@@ -3053,6 +3189,7 @@ bool Parser::ParseGenericFunctionExpression(ExprResult &Res) {
     return false;
   }
   else {
+    SmallVector<DeclRefExpr::GenericInstInfo::TypeArgument, 16> typeArgumentInfos;
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: Instead of 16, could you just use 4?   It is unlikely someone will write a function with more than 1 or 2  type parameters.

File: lib/Serialization/ASTReaderStmt.cpp [hunk 9]
```diff
@@ -599,8 +612,8 @@ void ASTStmtReader::VisitUnaryExprOrTypeTraitExpr(UnaryExprOrTypeTraitExpr *E) {
 
 void ASTStmtReader::VisitArraySubscriptExpr(ArraySubscriptExpr *E) {
   VisitExpr(E);
-  E->setLHS(Record.readSubExpr());
-  E->setRHS(Record.readSubExpr());
+  E->setLHS(Record.readSubExpr());
```
Comment (dtarditi) [inline lib/Serialization/ASTReaderStmt.cpp]: This looks like a spurious white-space change.  Could you revert it?

File: lib/Serialization/ASTReaderStmt.cpp [hunk 10]
```diff
@@ -662,7 +675,7 @@ void ASTStmtReader::VisitCastExpr(CastExpr *E) {
   VisitExpr(E);
   unsigned NumBaseSpecs = Record.readInt();
   assert(NumBaseSpecs == E->path_size());
-  E->setSubExpr(Record.readSubExpr());
+  E->setSubExpr(Record.readSubExpr());
```
Comment (dtarditi) [inline lib/Serialization/ASTReaderStmt.cpp]: This looks like a spurious white space change.  Could you revert it?

File: test/CheckedC/generic-function/generic-func-call-info.c (hunk 6 above)
Comment (dtarditi) [inline test/CheckedC/generic-function/generic-func-call-info.c]: Could you add tests for array types used as parameters?   There are 2 cases to cover:
- Incomplete array types as parameters:
`_For_any(T) int f(T[] arr)`
- Complete array types as parameters:
`_For_any(T) int f(T[10] arr)`

File: test/CheckedC/generic-function/generic-func-call-info.c [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+//
+// In CallExpr, the type of the DeclRefExpr should be completely instantiated,
+// by replacing all type variable types with specified type names in the list
+// of type names provided when calling generic function.
+//
+//
+// RUN: %clang_cc1 -ast-dump -fcheckedc-extension %s | FileCheck %s
+
+
+_For_any(T, S, R) T genericTest(T test, S steve, R ray) {
```
Comment (dtarditi) [inline test/CheckedC/generic-function/generic-func-call-info.c]: This test won't work after you make type variables be incomplete types.  Variables cannot have an incomplete type.

File: lib/Parse/ParseExpr.cpp [hunk 12]
```diff
@@ -3023,6 +3022,142 @@ ExprResult Parser::ParseBoundsExpression() {
...
+  case Type::TypeVariable: {
+    // Although Type Variable Type is wrapped with Typedef Type, there may be
+    // transformations in clang that eliminates Typedef.
+    const TypeVariableType *tvType = dyn_cast<TypeVariableType>(T);
+    if (!tvType) {
+      llvm_unreachable("Type cannot be empty when substituting type variable.");
+      return QT;
+    }
+    return typeNames[tvType->GetIndex()].typeName;
+  }
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: I think you also need to handle DecayedType and AdjustedType.   This will arise when a parameter is declared to have an array type or a function type.

File: lib/Parse/ParseExpr.cpp [hunk 13]
```diff
@@ -3023,6 +3022,142 @@ ExprResult Parser::ParseBoundsExpression() {
...
+// Recurse through types and substitute any TypedefType with TypeVariableType
+// as the underlying type.
+QualType Parser::SubstituteTypeVariable(QualType QT,
+  SmallVector<DeclRefExpr::GenericInstInfo::TypeArgument, 16> &typeNames) {
+  const Type *T = QT.getTypePtr();
+  switch (T->getTypeClass()) {
+  case Type::Pointer: {
+    const PointerType *pType = dyn_cast<PointerType>(T);
+    if (!pType) {
+      llvm_unreachable("Type cannot be empty when substituting type variable.");
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: I think a better unreachable error message would be "dynamic cast failed unexpectedly."

File: lib/Serialization/ASTReaderStmt.cpp (hunk 9 above)
Comment (jp1729) [inline lib/Serialization/ASTReaderStmt.cpp]: I made sure to save as Unix encoding, but it seems to have no effect.

File: lib/Serialization/ASTReaderStmt.cpp (hunk 10 above)
Comment (jp1729) [inline lib/Serialization/ASTReaderStmt.cpp]: I made sure to save as Unix encoding, but it seems to have no effect.

**Comment by jp1729** [comment]
It is ready for further review. I wrote the two test cases for array types in line 26 and line 30.
